frontend/src/utils/event.ts

```diff
@@ -0,0 +1,30 @@
+// a publish-subscribe pattern
+
+class Event {
+  eventList: Record<string, (...args: any[]) => any>;
+  constructor() {
+    this.eventList = {};
+  }
+
+  on(event: string, fn: (...args: any[]) => any) {
```

Comment:
let's define an ENUM type, otherwise it's very hard to track all events.

IIUC, event is more like type instead of the event iteself?